src/awkward/_broadcasting.py

```diff
@@ -1006,19 +945,96 @@ def continuation():
                     assert numoutputs == len(outcontents[-1])
-                numoutputs = len(outcontents[-1])
 
-            return tuple(
-                RecordArray(
-                    [x[i] for x in outcontents],
-                    None if istuple else fields,
-                    length,
-                    parameters=p,
-                )
-                for i, p in enumerate(parameters_factory(numoutputs))
+            assert numoutputs is not None
+
+        parameters = parameters_factory(numoutputs)
+        return tuple(
+            UnionArray.simplified(
+                Index8(tags),
+                Index64(index),
+                [x[i] for x in outcontents],
+                parameters=p,
             )
+            for i, p in enumerate(parameters)
+        )
+
+    def broadcast_any_indexed():
+        nextinputs = [x.project() if isinstance(x, IndexedArray) else x for x in inputs]
+        return apply_step(
+            backend,
+            nextinputs,
+            action,
+            depth,
+            copy.copy(depth_context),
+            lateral_context,
+            behavior,
+            options,
+        )
+
+    def broadcast_any_nd_numpy():
+        nextinputs = [
+            x.to_RegularArray() if isinstance(x, NumpyArray) else x for x in inputs
+        ]
+        return apply_step(
+            backend,
+            nextinputs,
+            action,
+            depth,
+            copy.copy(depth_context),
+            lateral_context,
+            behavior,
+            options,
+        )
+
+    def broadcast_any_unknown():
+        nextinputs = [
+            x.to_NumpyArray(np.float64, backend) if isinstance(x, EmptyArray) else x
+            for x in inputs
+        ]
+        return apply_step(
+            backend,
+            nextinputs,
+            action,
+            depth,
+            copy.copy(depth_context),
+            lateral_context,
+            behavior,
+            options,
+        )
+
+    def continuation():
+        # Any EmptyArrays?
+        if any(x.is_unknown for x in contents):
+            return broadcast_any_unknown()
+
+        # Any NumpyArrays with ndim != 1?
+        elif any(x.is_numpy and x.purelist_depth != 1 for x in contents):
+            return broadcast_any_nd_numpy()
+
+        # Any IndexedArrays?
+        elif any((x.is_indexed and not x.is_option) for x in contents):
+            return broadcast_any_indexed()
+
+        # Any UnionArrays?
+        elif any(x.is_union for x in contents):
+            return broadcast_any_union()
+
+        # Any option-types?
+        elif any(x.is_option for x in contents):
+            return broadcast_any_option()
+
+        # Any list-types?
+        elif any(x.is_list for x in contents):
+            return broadcast_any_list()
+
+        # Any RecordArrays?
+        elif any(x.is_record for x in contents):
+            return broadcast_any_record()
 
         else:
             raise ak._errors.wrap_error(
```

Comment:
In the end, this is the summary, the high-level overview of how broadcasting works. Splitting it out like this really helps to see this big picture! (And it's safer in that it limits variable scope.)